File: trials/matching_pair_sum.cpp
#include <iostream>
#include <vector>

using namespace std;
string matchingPairSum(vector<int>array);

int main(){
    vector<int> array = {1, 3, 7, 9};
    string answer = matchingPairSum(array);
    cout << answer << endl;
    return 0;
}


string matchingPairSum(vector<int>array) {
    int fronti = 0;
    int backi = array.size()-1;
    for (;backi!=fronti;){
        int sum = array[backi] + array[fronti];
        if (sum == 8){
            return "YES";
        }
        else if (sum > 8){
            backi --;
        } 
        else {
            fronti ++;
        }
    }
    return "NO";
}

File: trials/matching_pair_sum.py
def matching_pair_sum(array):
    """ """
    fronti = 0
    backi = len(array) - 1
    while fronti != backi:
        sum = array[fronti] + array[backi]
        if sum < 8:
            fronti += 1
        elif sum == 8:
            return "YES"
        else:
            backi -= 1
    return "NO"


if __name__ == "__main__":
    array = [1, 3, 4, 4]
    answer = matching_pair_sum(array)
    print("the answer", answer)
